engine/src/state_chain_observer/sc_observer/mod.rs

```diff
@@ -619,3 +612,163 @@ where
         Err(anyhow!("State Chain block stream ended"))
     }.instrument(info_span!("SCObserver")).boxed()).await
 }
+
+pub struct CeremonyIdCounters {
+	pub ethereum: CeremonyId,
+	pub polkadot: CeremonyId,
+	pub bitcoin: CeremonyId,
+}
+
+/// Get the ceremony id counters for each chain at the **start** of the given block without
+/// accessing the previous block
+pub async fn get_ceremony_id_counters_before_block<StateChainClient>(
+	block_hash: state_chain_runtime::Hash,
+	state_chain_client: Arc<StateChainClient>,
+) -> Result<CeremonyIdCounters, anyhow::Error>
+where
+	StateChainClient: SignedExtrinsicApi
+		+ UnsignedExtrinsicApi
+		+ 'static
+		+ Send
+		+ Sync
+		+ super::client::storage_api::StorageApi,
+{
+	// Get the vales of the ceremony id counters at the end of the block
+	let mut ceremony_id_counters = CeremonyIdCounters {
+		ethereum: state_chain_client
+			.storage_value::<pallet_cf_vaults::CeremonyIdCounter<
+				state_chain_runtime::Runtime,
+				state_chain_runtime::EthereumInstance,
+			>>(block_hash)
+			.await
+			.context("Failed to get Ethereum CeremonyIdCounter from SC")?,
+
+		polkadot: state_chain_client
+			.storage_value::<pallet_cf_vaults::CeremonyIdCounter<
+				state_chain_runtime::Runtime,
+				state_chain_runtime::PolkadotInstance,
+			>>(block_hash)
+			.await
+			.context("Failed to get Polkadot CeremonyIdCounter from SC")?,
+
+		bitcoin: state_chain_client
+			.storage_value::<pallet_cf_vaults::CeremonyIdCounter<
+				state_chain_runtime::Runtime,
+				state_chain_runtime::BitcoinInstance,
+			>>(block_hash)
+			.await
+			.context("Failed to get Bitcoin CeremonyIdCounter from SC")?,
+	};
+
+	// Reverse the counters by looking at the events in the block that would of incremented them
+	match state_chain_client
+		.storage_value::<frame_system::Events<state_chain_runtime::Runtime>>(block_hash)
+		.await
+	{
+		Ok(events) => {
+			for event_record in events {
+				match event_record.event {
+					state_chain_runtime::RuntimeEvent::EthereumVault(
+						pallet_cf_vaults::Event::KeygenRequest { .. },
+					) |
+					state_chain_runtime::RuntimeEvent::EthereumThresholdSigner(
+						pallet_cf_threshold_signature::Event::ThresholdSignatureRequest { .. },
+					) => {
+						ceremony_id_counters.ethereum -= 1;
```

Comment:
I think it would make more sense to simply look at the first event and sub one from its ceremony id (if not zero). And if there are no events get from the storage.